Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/ui/details/RecoveryCertificateDetailsFragment.kt

```diff
@@ -65,7 +68,65 @@ class RecoveryCertificateDetailsFragment : Fragment(R.layout.fragment_recovery_c
         certificate: RecoveryCertificate
     ) {
+        qrCodeCard.apply {
+            qrTitle.text = getString(R.string.recovery_certificate_name)
+            qrSubtitle.text = getString(
+                R.string.recovery_certificate_valid_until,
+                certificate.validUntil.toShortDayFormat()
+            )
+            when (certificate.getState()) {
+                is CwaCovidCertificate.State.ExpiringSoon -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(
+                        ContextCompat.getDrawable(
+                            requireContext(),
+                            R.drawable.ic_av_timer
+                        )
+                    )
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = getString(
+                        R.string.certificate_qr_expiration,
+                        certificate.headerExpiresAt.toShortDayFormat(),
+                        certificate.headerExpiresAt.toShortTimeFormat()
+                    )
+                    expirationStatusBody.visibility = View.VISIBLE
+                    expirationStatusBody.text = getText(R.string.expiration_info)
+                }
+
+                is CwaCovidCertificate.State.Expired -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(
+                        ContextCompat.getDrawable(
+                            requireContext(),
+                            R.drawable.ic_error_outline
+                        )
+                    )
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = getText(R.string.certificate_qr_expired)
+                    expirationStatusBody.visibility = View.VISIBLE
+                    expirationStatusBody.text = getText(R.string.expired_certificate_info)
+                    qrSubtitle.visibility = View.GONE
+                }
 
+                is CwaCovidCertificate.State.Invalid -> {
+                    expirationStatusIcon.visibility = View.VISIBLE
+                    expirationStatusIcon.setImageDrawable(
+                        ContextCompat.getDrawable(
+                            requireContext(),
+                            R.drawable.ic_error_outline
+                        )
+                    )
+                    expirationStatusText.visibility = View.VISIBLE
+                    expirationStatusText.text = getText(R.string.certificate_qr_invalid_signature)
+                    expirationStatusBody.visibility = View.VISIBLE
+                    expirationStatusBody.text = getText(R.string.invalid_certificate_signature_info)
+                    qrSubtitle.visibility = View.GONE
+                }
+
+                else -> {
+                }
+            }
+        }
```

Comment:
this code block seems to be common for all Certificates, should we create an extension function for that and reuse it for all?
```kotlin
private fun IncludeCertificateQrcodeCardBinding.bindValidityViews(certificate: CwaCovidCertificate) {
 // Add logic here 
}
```

Reply:
Yeah, it'd be nicer. I'm on it.